File: Tp1/Palindromo/PalindromoRecursivo.java


public class PalindromoRecursivo{
    public static void main(String[] args){

        String entrada = MyIO.readLine();
        int tamanho = entrada.length();
        boolean ehPalindromo;
        

        while(!strComp(entrada)){  //Repetir execucao do programa enquanto nao for escrito "FIM"

            ehPalindromo = ehPalindromo(entrada, tamanho-1, 0);  //Se a palavra tem 3 letras significa que ela vai de 0 ate 2

            escrever(ehPalindromo);

            entrada = MyIO.readLine();
            tamanho = entrada.length();
            
        }

    }


    public static boolean ehPalindromo(String frase, int tamanho, int pos){  //Funcao para testar se uma palavra eh palindromo usando recursividade
        boolean ehPalindromo = true;
        char letra1;
        char letra2;
        
        if(tamanho > pos){
            letra1 = frase.charAt(tamanho);
            letra2 = frase.charAt(pos);
            if(letra1 != letra2){
                ehPalindromo = false;
                tamanho = pos;
            }
            else{
                ehPalindromo = ehPalindromo(frase, tamanho-1, pos+1);
            }
        }

        return ehPalindromo;
    }


    public static void escrever(boolean ehPalindromo){  //Funcao para escrever na tela se eh palindromo
        if(ehPalindromo){
            MyIO.println("SIM");
        }
        else{
            MyIO.println("NAO");
        }
    }


    public static boolean strComp(String teste){  //Funcao para terminar a execucao do programa ao ser lido "FIM"
        boolean ehFim = true;
        String FIM = "FIM";

        for(int i = 0; i < 3; i++){
            if(FIM.charAt(i) != teste.charAt(i)){
                ehFim = false;
                i = 3;
            }
        }

        return ehFim;
    }
}

File: Tp1/Palindromo/palindromoRecursivo.c
#include <stdio.h>

#define bool short
#define true 1
#define false 0

//Prototipo das funcoes
int length(char frase[]);
bool ehPalindromo(char frase[], int tam, int pos);
bool strComp(char frase[]);


int main(){
    const int TAM = 1000;
    char entrada[TAM];

    fgets(entrada, TAM ,stdin);  //Ler valores
    int tam = length(entrada);  //Contar tamanho da frase recebida
    
    while(!strComp(entrada)){  //Repetir ate ler "FIM"
        if(ehPalindromo(entrada, length(entrada)-1, 0)){
            printf("SIM\n");
        }
        else{
            printf("NAO\n");
        }

        fgets(entrada, TAM ,stdin);
        tam = length(entrada);
    }

}


bool ehPalindromo(char frase[], int tam, int pos){  //Funcao recursiva para checar se uma frase eh um palindromo
    bool palindromo = true;
    
    if(tam > pos){
        if(frase[tam] != frase[pos]){
            palindromo = false;
            tam = pos;
        } else {
            palindromo = ehPalindromo(frase, tam-1,pos+1);
        }
    }

    return palindromo;
}




int length(char frase[]){  //Descobrir tamanho da frase
    int tam = 0;

    while(frase[tam] != '\0'){
        tam++;
    }

    tam--; //Retirando o '\0' da contagem

    return tam;
}


bool strComp(char frase[]){  //Comparador de strings com a palavra "FIM"
    bool ehFim = true;
    char FIM[] = "FIM";

    for(int i = 0; i < 3; i++){
        if(frase[i] != FIM[i]){
            ehFim = false;
            i = 3;
        }
    }

    return ehFim;
}